### PR TITLE
[interaction model] Remove map in InteractionModelEngine

### LINKED ISSUE #4275: Remove map from IM invoke command
Are we really OK to use std::map here? It's big and directly uses the heap, right? Believe we're planning to avoid the STL altogether.

_Originally posted by @mspang in https://github.com/project-chip/connectedhomeip/pull/4241#discussion_r552359944_

### PR DESCRIPTION
**Pull request by erjiaqing**
#### Problem

We are using map in IMEngine, however,

1. std::map is too heavy for mcu
2. We are unlikely to change handler of a command
3. The handler of commands are likely to be determined during compile time instead of run time.

 #### Summary of Changes

Introduce 

```
DispatchSingleClusterCommand(chip::ClusterId aClusterId, chip::CommandId aCommandId, chip::EndpointId aEndPointId, chip::GroupId aGroupId, chip::TLV::TLVReader & aReader, Command * apCommandObj)
```

For dispatch a single cluster command in CommandHandler / CommandSender.

 Fixes #4275

### CODE REVIEW

**Review by vivien-apple — APPROVED** [review]

**Review by yunhanw-google — APPROVED** [review]
Thanks!

**Review by andy31415 — APPROVED** [review]

**Review by selissia — APPROVED** [review]

File: src/app/InteractionModelEngine.cpp [hunk 1]
```diff
@@ -193,83 +194,17 @@ void InteractionModelEngine::OnResponseTimeout(Messaging::ExchangeContext * ec)
...
-                      "%d, Dispatcher = %p",
-                      aClusterId, aCommandId, aCommandRoleId, (void *) aDispatcher);
-    }
-
-    return err;
-}
-
-CHIP_ERROR InteractionModelEngine::DeregisterClusterCommandHandler(chip::ClusterId aClusterId, chip::CommandId aCommandId,
-                                                                   Command::CommandRoleId aCommandRoleId)
+void __attribute__((weak))
```
Comment (mspang) [inline src/app/InteractionModelEngine.cpp:197]: Why do we need this default implementation?

Comment (erjiaqing) [reply src/app/InteractionModelEngine.cpp:199]: Just to make linker happy before codegen for this is ready, added TODO to indicate that this should be removed.

**Review by woody-apple — APPROVED** [review]

**Comment by erjiaqing** [comment]
@jelderton @BroderickCarlin @saurabhst
Could you help review/approve this? thanks

**Comment by woody-apple** [comment]
@jelderton @saurabhst ?

**Comment by woody-apple** [comment]
@erjiaqing Looks like a conflict here, sorry :( can you please resolve?

**Review by saurabhst — APPROVED** [review]

**Comment by erjiaqing** [comment]
@woody-apple done.

**Pull request merged**